src/psyclone/psyir/transformations/arrayrange2loop_trans.py

```diff
@@ -214,6 +223,40 @@ def validate(self, node, options=None):
                             f"assignment '{node}'.")
                     break
 
+        if not options:
+            options = {}
+        allow_string_array = options.get("allow_string", False)
+        # If we allow string arrays then we can skip the check.
+        if allow_string_array:
+            return
+        lhs = node.lhs
+        # ArrayMixin datatype lookup can fail if the indices contain a
+        # Call or Intrinsic Call. We catch this exception and continue
+        # for now - TODO #1799
+        try:
+            if lhs.datatype.intrinsic == ScalarType.Intrinsic.CHARACTER:
+                raise TransformationError(
+                    "The ArrayRange2LoopTrans transformation doesn't allow "
+                    "character arrays by default. This can be enabled by "
+                    "passing the allow_string option to the transformation."
+                )
+        except NotImplementedError:
+            pass
+        for child in node.rhs.walk((Literal, Reference)):
+            # Skip over indices
+            if child.ancestor(ArrayReference) is not None:
+                continue
```

Comment:
The `test_validate_function_in_index` test shows why this is necessary. If you remove this line the code will fail. I think this was added because something in the exampes/transform stuff was failing because there was a Literal/Reference node in the rhs inside the array reference's we care about, e.g.
`b(:) = a(LBOUND(a):)`. I'm also unclear as to why this test doesn't fail anyway, since it should look at a and b but I assume because of whatever is wrong with the issue I say.

If you remove these lines then building in here fails:
`make[2]: Leaving directory '/home/aidan/PSyclone/examples/lfric/eg2'`

I think the example is why this code exists - I think because you can have unknownFortranType indexes or in Intrinsics and then it throws an exception.

Reply:
I think it fails on a different exception - not a NotImplementedError but one where either the .datatype or .intrinsic doesn't exist on the symbol.